### PR TITLE
Implement W3C Correlation Context propagator

### PR DESCRIPTION
**Pull request by sjkaris**
#124 

A couple open items on this PR.

1) I feel that the interface shouldn't take/return `[]core.KeyValue`, but was unsure what structure should be used in its place. `tag.Map` (soon to be renamed to `dctx.Map`/`distrubtedcontext.Map`) is a candidate, but doesn't allow for multiple same-valued keys as https://w3c.github.io/correlation-context/ states should be allowed. 

2) Properties defined in https://w3c.github.io/correlation-context/ are not handled very well, and its unclear to me what should be done with them. 

3) Do we want a separate interface for Correlation Context propagation vs SpanContext propagation

### CODE REVIEW

**Comment by freeformz** [comment]
What decides and when what keys/values need to be propagated ? Reading the linked doc I don't get a great feel beyond "user-defined baggage associated with the trace". Is that type of data likely to be different for each trace? Or is it something meant to be setup once or modified only irregularly?

**Comment by freeformz** [comment]
You could do something like this: https://github.com/sjkaris/opentelemetry-go/compare/feature/sk/w3s-correlation-ctx-prop...heroku:feature/sk/w3s-correlation-ctx-prop?expand=1

^ ends up conforming the new propagator to the propagation.TextFormatPropagator interface. To do so, I needed to expand core.SpanContext to also contain a []core.KeyValue. This has a side effect of making core.SpanContext not comparable, which can be dealt with in most cases, but not all. One example of a place that change affects is that the experimental streamer uses span contexts as a map key, which the new core.SpanContext can't be used as (adding a Hash() string function to spanContext or equivalent can get around that.

Thoughts?

**Comment by jmacd** [comment]
@sjkaris @freeformz 
I've read and thought about this a bit -- I'll admit I've been less focused on propagation issues than I should be. My understanding is that the grand "Context" consists of distributed context + span context. Whether they actually are stored in the same object or not seems to be the question here.  Is the expectation here that `Extract` will convert the `SpanContext` into standards-conventional key-values?  I believe until now, we've said that `Inject` should return the explicit key-values and SpanContext, not assemble a Golang `Context` directly--that should be up to the following span `Start`.

Am I understanding the issue, at least?

Also, @sjkaris I couldn't find what you mean by "doesn't allow for multiple same-valued keys"?

@freeformz About the experimental SDK.  Distributed context can't be "forgotten" by the streaming SDK the way all other state can. I believe we should maintain a `distributedcontext.Map` in the context for all SDKs, i.e., this is part of the API.  The experimental streaming back-end maintains the current state of the span using the SpanContext as the index key, which is independent from the current distributedcontext.Map values, and whenever the _Span_ changes it uses a new `EventID` to convey that the span is in a new state.  So if you added a new `M distributedcontext.Map` field in SpanContext to store distributedcontext.Map, that would be fine, but it wouldn't be used to index current state in the streaming SDK.

**Comment by sjkaris** [comment]
@jmacd, yep I think you understand the issue. Just to re-state, from what I gathered at the SIG meeting, this PR should be changed such that we return both the `SpanContext` and the `dctx.Map` in `Extract` and additionally take a `dctx.Map` as an argument to `Inject` to pass arbitrary keys along. 

Its poorly worded, but what I meant by "doesn't allow for multiple same-valued keys", is basically that the `dctx.Map` doesn't let you have multiple entries for the same key, like a slice or list would allow. I'm not sure that this is actually problematic.

**Comment by sjkaris** [comment]
I would be interested to hear @freeformz thoughts on this as well though

**Comment by freeformz** [comment]
I don't know what happened at the SIG meeting unfortunately. I'd prefer to not to add another interface like the original PR did, but make an interface that works for all of these types of headers, possibly with optional secondary interfaces as needed.

I am a bit worried about the question of needing to support multiple values for the same key. If that is a requirement then dctx.Map will need to be changed or can't be used for this.

**Comment by rghetia** [comment]
Separation of correlation context from observability is discussed [here](https://github.com/open-telemetry/oteps/pulls). As part of this discussion, the interface for propagators will change as well. May I suggest that we finish this PR with a single propagator that handles both, trace-context and distributed-context with following signature?
```go
inject(ctx context.Context, supplier Supplier)
extract(ctx, contex.Context, supplier Supplier) (SpanContext, dctx.Map)
```

Regarding multiple values correlation context - As per the [DistributedContext spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-distributedcontext.md), a key can only have one value. So for now, if there are duplicate keys during extraction we either take the last or the first entry. Open an issue to resolve the multiple value support in the spec.

**Comment by freeformz** [comment]
Separation discussion URL: https://github.com/open-telemetry/oteps/pull/42

**Comment by rghetia** [comment]
> Separation discussion URL: [open-telemetry/oteps#42](https://github.com/open-telemetry/oteps/pull/42)

thanks for posting the correct link.

File: api/propagation/propagator.go [hunk 1]
```diff
@@ -18,22 +18,47 @@ import (
...
+	// * allow pre-allocation of fields, especially in systems like gRPC Metadata
+	// * allow a single-pass over an iterator (ex OpenTracing has no getter in TextMap)
+	GetAllKeys() []string
+}
+
+// TextFormatBaggagePropagator is an interface that specifies methods to inject and extract
+// DistributedContext into/from a carrier using Supplier interface.
+// For example, HTTP Correlation Context propagator would encode DistributedContext into W3C Trace
+// Correlation Context Header and set the header into HttpRequest.
+type TextFormatCorrelationContextPropagator interface {
```
Comment (rghetia) [inline api/propagation/propagator.go]: Is this a Baggage or CorrelationContext? Either way it should be removed. If it is baggage then we want to hold on to adding this until we have clarity on whether to support baggage as part of open-telemetery or an independent contrib package. If it is correlation-context (for now distributedContext) then it is already part of trace-context propagator.

File: api/propagation/propagator.go [hunk 2]
```diff
@@ -18,22 +18,47 @@ import (
...
 // For example, HTTP Trace Context propagator would encode SpanContext into W3C Trace
 // Context Header and set the header into HttpRequest.
 type TextFormatPropagator interface {
 	// Inject method retrieves current SpanContext from the ctx, encodes it into propagator
 	// specific format and then injects the encoded SpanContext using supplier into a carrier
-	// associated with the supplier.
-	Inject(ctx context.Context, supplier Supplier)
+	// associated with the supplier. It also takes a correlationCtx whose values will be
+	// injected into a carrier using the supplier.
+	Inject(ctx context.Context, correlationCtx dctx.Map, supplier Supplier)
```
Comment (rghetia) [inline api/propagation/propagator.go]: dctx should already be in ctx. So Inject should simply retrieve it from the context.

File: plugin/httptrace/httptrace.go [hunk 3]
```diff
@@ -36,16 +37,24 @@ var (
...
+	sc, correlationCtx := propagator.Extract(ctx, req.Header)
 
 	attrs := []core.KeyValue{
 		URLKey.String(req.URL.String()),
 		// Etc.
 	}
+	correlationCtx.Foreach(func(kv core.KeyValue) bool {
+		attrs = append(attrs, kv)
+		return true
+	})
```
Comment (rghetia) [inline plugin/httptrace/httptrace.go]: should this be appended to attributes or returned second argument. I believe 'Context Entries' refer to DistributedContext here. @jmacd , can you please confirm?

Comment (sjkaris) [reply plugin/httptrace/httptrace.go]: Ah, I suspect you are right, this should be returned as the second arg.

File: api/propagation/propagator.go (hunk 1 above)
Comment (sjkaris) [reply api/propagation/propagator.go]: Oops, I didn't realize this didn't get deleted. This is unused and should be removed.

File: plugin/httptrace/httptrace.go (hunk 3 above)
Comment (jmacd) [reply plugin/httptrace/httptrace.go]: (I'm open to improvements on this API.)

**Comment by rghetia** [comment]
> @sjkaris, LGTM. Can you please rebase?

one more conflict due to #218

**Pull request merged**